Services/Component/classes/Setup/class.ilComponentBuildPluginInfoObjective.php

```diff
@@ -0,0 +1,99 @@
+<?php
+
+use ILIAS\Setup;
+
+class ilComponentBuildPluginInfoObjective extends Setup\Artifact\BuildArtifactObjective
+{
+    const BASE_PATH = "./Customizing/global/plugins/";
+    const PLUGIN_PHP = "plugin.php";
+
+    public function getArtifactPath() : string
+    {
+        return \ilArtifactComponentDataDB::PLUGIN_DATA_PATH;
+    }
+
+
+    public function build() : Setup\Artifact
+    {
+        $data = [];
+        foreach (["Modules", "Services"] as $type) {
+            $components = $this->scanDir(static::BASE_PATH . "$type");
+            foreach ($components as $component) {
+                $slots = $this->scanDir(static::BASE_PATH . "$type/$component");
+                foreach ($slots as $slot) {
+                    $plugins = $this->scanDir(static::BASE_PATH . "$type/$component/$slot");
+                    foreach ($plugins as $plugin) {
+                        $this->addPlugin($data, $type, $component, $slot, $plugin);
+                    }
+                }
+            }
+        }
+        return new Setup\Artifact\ArrayArtifact($data);
+    }
+
+    protected function addPlugin(array &$data, string $type, string $component, string $slot, string $plugin) : void
+    {
+        $path = static::BASE_PATH . "$type/$component/$slot/$plugin/" . static::PLUGIN_PHP;
+        $file = $this->readFile($path);
+        if (is_null($file)) {
+            throw new \RuntimeException(
+                "Cannot read $path."
+            );
+        }
+
+        eval("?>" . $file);
```

Comment:
Just a short inline feedback: Do we really want to rely on `eval` here?

Reply:
I did not like this too. But: Any better idea? Also: Is this really a problem?

Reply:
>  Any better idea?

ATM not, therefore I had to competely understand the change (mental model) ;-).

> Is this really a problem?

I don't think so. In the past some ILIAS institutions blocked the `eval` function (because of security concerns) and only allowed the execution during the setup. So with your database update step revision we got finally rid of `eval`, and now we introduce it once again :-).

Reply:
Yeah, I'm really looking to get rid of this completely soon =) This is what I want to talk about (among other things) on the Project JF.

Reply:
Why dont you use an `include`? Like it was used in `ilPluginAdmin`:

```
 protected function parsePluginPhp($plugin_php_file)
    {
        include($plugin_php_file);

        $values = [
            "version" => $version,
            "id" => $id,
            "ilias_min_version" => $ilias_min_version,
            "ilias_max_version" => $ilias_max_version,
            "responsible" => $responsible,
            "responsible_mail" => $responsible_mail,
            "learning_progress" => (bool) ($learning_progress ?? false),
            "supports_export" => (bool) ($supports_export ?? false),
            "supports_cli_setup" => (bool) ($supports_cli_setup ?? true)
        ];

        return $values;
    }
```

Is this faster?

Reply:
@PurHur: The tests indeed.